### PR TITLE
Sylius resource type condition

### PR DESCRIPTION
**Pull request by RandyCupic**
In one of the projects we have a need to be able to map layout to all instances of a specific Sylius resource (product, taxon as well as any other eventual custom resources that are Sylius resources), in order to be able to build those pages (eg. product page) with layouts.

This PR introduces a new condition type that enables you to map a layout to all Sylius products or taxons. The list of available resources is done through semantic configuration so it can be easily extended in project to add new custom Sylius resources. Furthermore, Sylius event dispatcher has been decorated to dispatch generic events (eg. `sylius_resource_show`) and then we have a listener which resolves the resource and adds it to request so matching should be done automatically. Eg. if you create a new custom Sylius resource in your project, you only need to add it to the semantic configuration. Matching will automatically work.

Here's an example of semantic configuration (default one):

```
netgen_layouts_sylius:
    resource_type_condition:
        available_resources:
            Sylius\Component\Product\Model\ProductInterface: product
            Sylius\Component\Taxonomy\Model\TaxonInterface: taxon
```

### CODE REVIEW

File: bundle/EventDispatcher/SyliusEventDispatcher.php [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+use Sylius\Bundle\ResourceBundle\Event\ResourceControllerEvent;
+use Sylius\Component\Resource\Model\ResourceInterface;
+use Symfony\Contracts\EventDispatcher\EventDispatcherInterface as SymfonyEventDispatcherInterface;
+
+use function sprintf;
+
+final class SyliusEventDispatcher implements EventDispatcherInterface
+{
+    public function __construct(
+        private readonly EventDispatcherInterface $innerEventDispatcher,
```
Comment (emodric) [inline bundle/EventDispatcher/SyliusEventDispatcher.php]: Do not use `private readonly` please, only `private`.

File: lib/Layout/Resolver/Form/ConditionType/Mapper/ResourceType.php [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+    /**
+     * @return array<string, string>
+     */
+    private function getResourceTypeList(): array
+    {
+        $resourceTypeList = [];
+
+        foreach ($this->allowedResources as $type) {
+            $resourceTypeList[$this->humanizeType($type)] = $type;
```
Comment (emodric) [inline lib/Layout/Resolver/Form/ConditionType/Mapper/ResourceType.php]: This should go through `choice_label` option, e.g.

```
'choices' => $this->allowedResources,
'choice_label' => fn (string $type): string => $this->humanizeType($type),
```

File: tests/bundle/EventListener/Admin/MainMenuBuilderListenerTest.php [hunk 3]
```diff
@@ -22,7 +22,7 @@ final class MainMenuBuilderListenerTest extends TestCase
 {
     private MainMenuBuilderListener $listener;
 
-    private MockObject&AuthorizationCheckerInterface $authCheckerMock;
+    private AuthorizationCheckerInterface&MockObject $authCheckerMock;
```
Comment (emodric) [inline tests/bundle/EventListener/Admin/MainMenuBuilderListenerTest.php:25]: ~Is this done by CS fixer? If not, please do not turn this around. If it is, how did `SyliusEventDispatcherTest` manage to have it different?~

Argh, sorted by alphabet.

File: tests/bundle/EventListener/Shop/ProductIndexListenerTest.php [hunk 4]
```diff
@@ -23,7 +23,7 @@ final class ProductIndexListenerTest extends TestCase
 
     private MockObject&TaxonRepositoryInterface $taxonRepositoryMock;
 
-    private MockObject&LocaleContextInterface $localeContextMock;
+    private LocaleContextInterface&MockObject $localeContextMock;
```
Comment (emodric) [inline tests/bundle/EventListener/Shop/ProductIndexListenerTest.php:26]: Same remark as with `MainMenuBuilderListenerTest.php`

File: tests/lib/Layout/Resolver/Form/ConditionType/Mapper/ResourceTypeTest.php [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+#[CoversClass(ResourceType::class)]
+final class ResourceTypeTest extends TestCase
+{
+    private ResourceType $mapper;
+
+    protected function setUp(): void
+    {
+        $allowedResources = [
+            'Sylius\Component\Product\Model\ProductInterface' => 'product',
```
Comment (emodric) [inline tests/lib/Layout/Resolver/Form/ConditionType/Mapper/ResourceTypeTest.php]: Use `::class` constants here, please.

File: tests/lib/Validator/ResourceTypeValidatorTest.php [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    {
+        parent::setUp();
+
+        $this->constraint = new ResourceType();
+    }
+
+    public function getValidator(): ConstraintValidatorInterface
+    {
+        $allowedResources = [
+            'Sylius\Component\Product\Model\ProductInterface' => 'product',
```
Comment (emodric) [inline tests/lib/Validator/ResourceTypeValidatorTest.php]: `::class` constants.

File: tests/lib/Layout/Resolver/ConditionType/ResourceTypeTest.php [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+final class ResourceTypeTest extends TestCase
+{
+    private ResourceType $conditionType;
+
+    private ValidatorInterface $validator;
+
+    protected function setUp(): void
+    {
+        $allowedResources = [
+            'Sylius\Component\Product\Model\ProductInterface' => 'product',
```
Comment (emodric) [inline tests/lib/Layout/Resolver/ConditionType/ResourceTypeTest.php]: `::class` constants here too.

File: bundle/EventDispatcher/SyliusEventDispatcher.php (hunk 1 above)
Comment (RandyCupic) [reply bundle/EventDispatcher/SyliusEventDispatcher.php]: Fixed in https://github.com/netgen-layouts/layouts-sylius/pull/11/commits/93ab79cc5e0f027f9222331aa0551e4358fa80db

File: lib/Layout/Resolver/Form/ConditionType/Mapper/ResourceType.php (hunk 2 above)
Comment (RandyCupic) [reply lib/Layout/Resolver/Form/ConditionType/Mapper/ResourceType.php]: Fixed in https://github.com/netgen-layouts/layouts-sylius/pull/11/commits/93ab79cc5e0f027f9222331aa0551e4358fa80db

File: tests/bundle/EventListener/Admin/MainMenuBuilderListenerTest.php (hunk 3 above)
Comment (RandyCupic) [reply tests/bundle/EventListener/Admin/MainMenuBuilderListenerTest.php:25]: Yep :)

File: tests/lib/Layout/Resolver/Form/ConditionType/Mapper/ResourceTypeTest.php (hunk 5 above)
Comment (RandyCupic) [reply tests/lib/Layout/Resolver/Form/ConditionType/Mapper/ResourceTypeTest.php]: Fixed in https://github.com/netgen-layouts/layouts-sylius/pull/11/commits/93ab79cc5e0f027f9222331aa0551e4358fa80db

File: tests/lib/Validator/ResourceTypeValidatorTest.php (hunk 6 above)
Comment (RandyCupic) [reply tests/lib/Validator/ResourceTypeValidatorTest.php]: Fixed in https://github.com/netgen-layouts/layouts-sylius/pull/11/commits/93ab79cc5e0f027f9222331aa0551e4358fa80db

File: tests/lib/Layout/Resolver/ConditionType/ResourceTypeTest.php (hunk 7 above)
Comment (RandyCupic) [reply tests/lib/Layout/Resolver/ConditionType/ResourceTypeTest.php]: Fixed in https://github.com/netgen-layouts/layouts-sylius/pull/11/commits/93ab79cc5e0f027f9222331aa0551e4358fa80db

**Comment by emodric** [comment]
Thanks @RandyCupic !

**Pull request merged**